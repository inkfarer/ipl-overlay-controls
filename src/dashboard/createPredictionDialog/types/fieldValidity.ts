export interface FieldValidity{
    [key: string]: boolean
}