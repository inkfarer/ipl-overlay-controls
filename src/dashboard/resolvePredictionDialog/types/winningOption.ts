export interface WinningOption{
    validOption: boolean;
    optionIndex: number;
    optionTitle: string;
}