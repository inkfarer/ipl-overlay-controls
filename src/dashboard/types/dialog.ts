export type NodecgDialog = Element & { open: () => void, close: () => void };
