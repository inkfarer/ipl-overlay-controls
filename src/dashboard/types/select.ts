export type SelectOptions = Array<{ name: string, value: string, disabled?: boolean }>;

export type SelectOptionGroups = Array<{ name: string, options: Array<SelectOptions> }>;
