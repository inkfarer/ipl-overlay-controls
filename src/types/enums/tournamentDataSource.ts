export enum TournamentDataSource {
    BATTLEFY = 'BATTLEFY',
    SMASHGG = 'SMASHGG',
    UPLOAD = 'UPLOAD',
    SENDOU_INK = 'SENDOU_INK',
    UNKNOWN = 'UNKNOWN'
}

export class TournamentDataSourceHelper {
    static toPrettyString(value: TournamentDataSource | `${TournamentDataSource}`): string {
        return {
            [TournamentDataSource.BATTLEFY]: 'Battlefy',
            [TournamentDataSource.SMASHGG]: 'Smash.gg',
            [TournamentDataSource.UPLOAD]: 'Uploaded file',
            [TournamentDataSource.SENDOU_INK]: 'sendou.ink',
            [TournamentDataSource.UNKNOWN]: 'Unknown'
        }[value];
    }
}
