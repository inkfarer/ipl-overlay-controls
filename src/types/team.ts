import { Player } from 'types/player';

export interface Team {
    id: string;
    name: string;
    logoUrl?: string;
    players: Player[];
}
