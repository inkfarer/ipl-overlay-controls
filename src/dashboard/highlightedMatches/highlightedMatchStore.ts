import { NodeCGBrowser } from 'nodecg/browser';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import {
    HighlightedMatches,
    TournamentData
} from 'schemas';
import { SelectOptions } from '../types/select';
import { TournamentDataSource } from 'types/enums/tournamentDataSource';

const highlightedMatches = nodecg.Replicant<HighlightedMatches>('highlightedMatches');
const tournamentData = nodecg.Replicant<TournamentData>('tournamentData');

export const highlightedMatchReps = [ highlightedMatches, tournamentData ];

export interface HighlightedMatchStore {
    highlightedMatches: HighlightedMatches;
    tournamentData: TournamentData;
}

export const highlightedMatchStore = createStore<HighlightedMatchStore>({
    state: {
        highlightedMatches: null,
        tournamentData: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        }
    },
    actions: {
        async getHighlightedMatches(store, { options }: { options: SelectOptions }): Promise<void> {
            const getAllMatches = options.some(option => option.value === 'all');

            if (getAllMatches) {
                return nodecg.sendMessage('getHighlightedMatches', { getAllMatches });
            } else {
                const values = options.map(option => option.value);

                switch (store.state.tournamentData.meta.source) {
                    case TournamentDataSource.BATTLEFY:
                        return nodecg.sendMessage('getHighlightedMatches', {
                            getAllMatches: false,
                            stages: values
                        });
                    case TournamentDataSource.SMASHGG:
                        return nodecg.sendMessage('getHighlightedMatches', {
                            getAllMatches: false,
                            streamIDs: values.map(value => parseInt(value))
                        });
                    default:
                        throw new Error(`Cannot import data from source '${store.state.tournamentData.meta.source}'`);
                }
            }
        },
        async setNextMatch(store, { teamAId, teamBId }: { teamAId: string, teamBId: string }): Promise<void> {
            return nodecg.sendMessage('setNextRound', { teamAId, teamBId });
        }
    }
});

export const highlightedMatchStoreKey: InjectionKey<Store<HighlightedMatchStore>> = Symbol();

export function useHighlightedMatchStore(): Store<HighlightedMatchStore> {
    return useStore(highlightedMatchStoreKey);
}
