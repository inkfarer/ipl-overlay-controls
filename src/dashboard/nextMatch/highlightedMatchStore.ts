import { NodeCGBrowser } from 'nodecg/browser';
import { HighlightedMatches, TournamentData } from 'schemas';
import { SelectOptions } from '../types/select';
import { TournamentDataSource } from 'types/enums/tournamentDataSource';
import { SetNextRoundRequest } from 'types/messages/rounds';
import { defineStore } from 'pinia';

const highlightedMatches = nodecg.Replicant<HighlightedMatches>('highlightedMatches');
const tournamentData = nodecg.Replicant<TournamentData>('tournamentData');

export const highlightedMatchReps = [ highlightedMatches, tournamentData ];

export interface HighlightedMatchStore {
    highlightedMatches: HighlightedMatches;
    tournamentData: TournamentData;
}

export const useHighlightedMatchStore = defineStore('highlightedMatches', {
    state: () => ({
        highlightedMatches: null,
        tournamentData: null
    } as HighlightedMatchStore),
    actions: {
        async getHighlightedMatches({ options }: { options: SelectOptions }): Promise<void> {
            const getAllMatches = options.some(option => option.value === 'all');

            if (getAllMatches) {
                return nodecg.sendMessage('getHighlightedMatches', { getAllMatches });
            } else {
                const values = options.map(option => option.value);

                switch (this.tournamentData.meta.source) {
                    case TournamentDataSource.BATTLEFY:
                        return nodecg.sendMessage('getHighlightedMatches', {
                            getAllMatches: false,
                            stages: values
                        });
                    case TournamentDataSource.SMASHGG:
                        return nodecg.sendMessage('getHighlightedMatches', {
                            getAllMatches: false,
                            streamIDs: values.map(value => parseInt(value))
                        });
                    default:
                        throw new Error(`Cannot import data from source '${this.tournamentData.meta.source}'`);
                }
            }
        },
        async setNextMatch(data: SetNextRoundRequest): Promise<void> {
            return nodecg.sendMessage('setNextRound', data);
        }
    }
});
