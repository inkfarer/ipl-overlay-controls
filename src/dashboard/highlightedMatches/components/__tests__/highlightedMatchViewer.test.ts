import HighlightedMatchViewer from '../highlightedMatchViewer.vue';
import { createStore } from 'vuex';
import { HighlightedMatchStore, highlightedMatchStoreKey } from '../../highlightedMatchStore';
import { TournamentDataSource } from 'types/enums/tournamentDataSource';
import { config, mount } from '@vue/test-utils';
import { Team } from 'types/team';

describe('HighlightedMatchViewer', () => {
    config.global.stubs = {
        IplSelect: true,
        IplDataRow: true,
        IplButton: true
    };

    const mockSetNextMatch = jest.fn();
    const mockTeam: Team = { id: '1234', name: 'mock team', showLogo: false, players: []};

    function createHighlightedMatchStore() {
        return createStore<HighlightedMatchStore>({
            state: {
                tournamentData: {
                    meta: {
                        id: 'tournament123',
                        source: TournamentDataSource.UNKNOWN
                    },
                    teams: []
                },
                highlightedMatches: null
            },
            actions: {
                setNextMatch: mockSetNextMatch
            }
        });
    }

    it('matches snapshot when no matches are loaded', () => {
        const store = createHighlightedMatchStore();
        store.state.highlightedMatches = [];
        const wrapper = mount(HighlightedMatchViewer, {
            global: {
                plugins: [[store, highlightedMatchStoreKey]]
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('matches snapshot', () => {
        const store = createHighlightedMatchStore();
        store.state.highlightedMatches = [
            { meta: { name: 'cool match', id: '1234' }, teamA: mockTeam, teamB: mockTeam },
            { meta: { name: 'cooler match', id: '567' }, teamA: mockTeam, teamB: mockTeam }
        ];
        const wrapper = mount(HighlightedMatchViewer, {
            global: {
                plugins: [[store, highlightedMatchStoreKey]]
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('selects first match when highlighted matches are updated', async () => {
        const store = createHighlightedMatchStore();
        store.state.highlightedMatches = [
            { meta: { name: 'cool match', id: '1234' }, teamA: mockTeam, teamB: mockTeam },
            { meta: { name: 'cooler match', id: '567' }, teamA: mockTeam, teamB: mockTeam }
        ];
        const wrapper = mount(HighlightedMatchViewer, {
            global: {
                plugins: [[store, highlightedMatchStoreKey]]
            }
        });

        expect(wrapper.getComponent('[data-test="match-selector"]').attributes().modelvalue).toEqual('1234');
        store.state.highlightedMatches = [
            { meta: { name: 'new match', id: '678' }, teamA: mockTeam, teamB: mockTeam },
            { meta: { name: 'new match 2', id: '3089754' }, teamA: mockTeam, teamB: mockTeam }
        ];
        await wrapper.vm.$nextTick();
        expect(wrapper.getComponent('[data-test="match-selector"]').attributes().modelvalue).toEqual('678');
    });

    it('displays team names for selected match', async () => {
        const store = createHighlightedMatchStore();
        store.state.highlightedMatches = [
            { meta: { name: 'cool match', id: '1234' }, teamA: mockTeam, teamB: mockTeam },
            {
                meta: { name: 'cooler match', id: '567' },
                teamA: {
                    ...mockTeam,
                    name: 'mock team a'
                },
                teamB: {
                    ...mockTeam,
                    name: 'mock team b'
                }
            }
        ];
        const wrapper = mount(HighlightedMatchViewer, {
            global: {
                plugins: [[store, highlightedMatchStoreKey]]
            }
        });

        wrapper.getComponent('[data-test="match-selector"]').vm.$emit('update:modelValue', '567');
        await wrapper.vm.$nextTick();
        expect(wrapper.getComponent('[data-test="team-a-name-display"]').attributes().value).toEqual('mock team a');
        expect(wrapper.getComponent('[data-test="team-b-name-display"]').attributes().value).toEqual('mock team b');
    });

    it('commits to store on set next match button press', async () => {
        const store = createHighlightedMatchStore();
        store.state.highlightedMatches = [
            {
                meta: { name: 'cooler match', id: '567' },
                teamA: {
                    ...mockTeam,
                    id: '1234'
                },
                teamB: {
                    ...mockTeam,
                    id: '5678'
                }
            }
        ];
        const wrapper = mount(HighlightedMatchViewer, {
            global: {
                plugins: [[store, highlightedMatchStoreKey]]
            }
        });

        wrapper.getComponent('[data-test="set-next-match-button"]').vm.$emit('click');

        expect(mockSetNextMatch).toHaveBeenCalledWith(expect.any(Object), {
            teamAId: '1234',
            teamBId: '5678'
        });
    });

    it('disables set next match button if selected match data is not found', async () => {
        const store = createHighlightedMatchStore();
        store.state.highlightedMatches = [
            {
                meta: { name: 'cooler match', id: '567' },
                teamA: mockTeam,
                teamB: mockTeam
            }
        ];
        const wrapper = mount(HighlightedMatchViewer, {
            global: {
                plugins: [[store, highlightedMatchStoreKey]]
            }
        });

        wrapper.getComponent('[data-test="match-selector"]').vm.$emit('update:modelValue', 'something');
        await wrapper.vm.$nextTick();

        expect(wrapper.getComponent('[data-test="set-next-match-button"]').attributes().disabled).toEqual('true');
    });
});
