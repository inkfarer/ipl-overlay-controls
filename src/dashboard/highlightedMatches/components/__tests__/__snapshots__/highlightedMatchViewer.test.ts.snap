// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighlightedMatchViewer matches snapshot 1`] = `
"<div class=\\"ipl-space color-dark\\" data-v-11398c0e=\\"\\">
  <ipl-select-stub modelvalue=\\"1234\\" label=\\"Match\\" options=\\"[object Object],[object Object]\\" disabled=\\"false\\" data-test=\\"match-selector\\"></ipl-select-stub>
  <ipl-data-row-stub label=\\"Team A\\" data-test=\\"team-a-name-display\\" value=\\"mock team\\"></ipl-data-row-stub>
  <ipl-data-row-stub label=\\"Team B\\" data-test=\\"team-b-name-display\\" value=\\"mock team\\"></ipl-data-row-stub>
  <ipl-button-stub label=\\"Set next match\\" color=\\"blue\\" disabled=\\"false\\" async=\\"false\\" progressmessage=\\"Working...\\" successmessage=\\"Done!\\" requiresconfirmation=\\"false\\" shortconfirmationmessage=\\"false\\" small=\\"false\\" disableonsuccess=\\"false\\" class=\\"m-t-8\\" data-test=\\"set-next-match-button\\"></ipl-button-stub>
</div>"
`;

exports[`HighlightedMatchViewer matches snapshot when no matches are loaded 1`] = `
"<div class=\\"ipl-message__wrapper layout horizontal center-vertical ipl-message__type-info\\" data-v-13b37b8d=\\"\\"><svg class=\\"svg-inline--fa fa-info-circle fa-w-16 icon\\" aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"info-circle\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\" data-v-13b37b8d=\\"\\">
    <path class=\\"\\" fill=\\"currentColor\\" d=\\"M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z\\"></path>
  </svg>
  <div class=\\"message-content\\" data-v-13b37b8d=\\"\\">
    <!----> No matches loaded.
  </div>
</div>"
`;
